### PR TITLE
Upgrade prng to c++

### PR DESCRIPTION
**Pull request by fireice-uk**
Use C++ and remove `INITIALIZER`

I removed a global random seed, as this is insecure. Instead we either use the randomness function directly or we use keccak hashes of a random seed stored on stack.

This might complicate test slightly. If you need to fix the prng for the test, I recommend wiring something based on C++11 mt just for tests, enabled by `#define`.

### CODE REVIEW

File: src/crypto/random.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	{
+		std::cerr << "Exit Failure :: open /dev/urandom" << std::endl;
+		std::abort();
+	}
+#endif
+}
+
+void prng::generate_random(uint8_t* output, size_t size_bytes)
+{
+	if(hnd == nullptr)
```
Comment (psychocrypt) [inline src/crypto/random.cpp]: This is not thread save and should make issues in an parallel environment. This will create a race condition because `hnd` is updated in `start`
If this is called in the constructor of the singleton than it is by definition thread save.

File: src/crypto/random.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	}
+#endif
+	delete hnd;
+}
+
+void prng::start()
+{
+	hnd = new prng_handle;
+#if defined(_WIN32)
+	if(!CryptAcquireContext(&hnd->prov, NULL, NULL, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT))
```
Comment (psychocrypt) [inline src/crypto/random.cpp]: Why not using standard c++ and use https://en.cppreference.com/w/cpp/numeric/random/random_device/random_device
This will avoid that we need to fiddle around with platform specific methods to access the random number generator.

File: src/crypto/random.hpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+public:
+	prng() {}
+	~prng();
+
+	void generate_random(uint8_t* output, size_t size_bytes);
+	inline void generate_random(void* output, size_t size_bytes) { generate_random(reinterpret_cast<uint8_t*>(output), size_bytes); }
+
+	inline static prng& inst()
+	{
+		static thread_local prng inst;
```
Comment (psychocrypt) [inline src/crypto/random.hpp]: Could it be that this class should not be thread save and have a instant per thread. Please add a doxygen doc sting to the class and describe the behavior e.g. thread safety and other edge cases

File: tests/crypto/random.c [hunk 4]
```diff
@@ -28,11 +28,11 @@
...
 
-#include "crypto/random.c"
+//#include "crypto/random.c"
 
 #include "crypto-tests.h"
 
 void setup_random(void)
 {
-	memset(&state, 42, sizeof(union hash_state));
+	//memset(&state, 42, sizeof(union hash_state));
```
Comment (psychocrypt) [inline tests/crypto/random.c]: Please fix the test cases, else it is not possible to merge.

File: src/crypto/random.cpp (hunk 1 above)
Comment (fireice-uk) [reply src/crypto/random.cpp]: Whole singleton is `thread_local`, we don't need to worry about this. This is an explicit optimisation as `thread_local` variables might be created on the thread start.

File: src/crypto/random.cpp (hunk 2 above)
Comment (fireice-uk) [reply src/crypto/random.cpp]: This is security critical. We won't use anything not crypto-grade.

File: src/crypto/random.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+		std::cerr << "Exit Failure :: close /dev/urandom " << std::endl; 
+		std::abort();
+	}
+#endif
+	delete hnd;
+}
+
+void prng::start()
+{
+	hnd = new prng_handle;
```
Comment (psychocrypt) [inline src/crypto/random.cpp]: please check that `new` is not returning a `nullptr`

File: src/crypto/random.cpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		std::abort();
+	}
+#else
+	if(close(hnd->fd) < 0)
+	{
+		std::cerr << "Exit Failure :: close /dev/urandom " << std::endl; 
+		std::abort();
+	}
+#endif
+	delete hnd;
```
Comment (psychocrypt) [inline src/crypto/random.cpp]: if we use `std::unique_ptr` this is not possible

File: src/crypto/random.cpp (hunk 5 above)
Comment (fireice-uk) [reply src/crypto/random.cpp]: http://www.cplusplus.com/reference/new/bad_alloc/

But honestly, treating out-of-memory is like treating a bullet wound to the head. What are you gonna do? Print an error message? Good luck!

File: src/crypto/random.cpp (hunk 6 above)
Comment (fireice-uk) [reply src/crypto/random.cpp]: I was a bit split here too, but in the end decided against since there is only one possible deallocation.

File: src/crypto/random.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+	}
+	else
+	{
+		uint64_t buffer[5];
+		buffer[0] = 0;
+		generate_system_random_bytes(reinterpret_cast<uint8_t*>(buffer+1), sizeof(buffer) - sizeof(uint64_t));
+
+		while(size_bytes > 200)
+		{
+			buffer[0]++;
```
Comment (psychocrypt) [inline src/crypto/random.cpp]: only to be sure, it is not a problem that `buffer[0]` always start with `0`

Comment (fireice-uk) [reply src/crypto/random.cpp]: This is correct pseudocode for this system would be:

`random bytes = H(1 || seed) || H(2 || seed) || H(3 || seed) || ...` until we have the needed number of random bytes.
